### PR TITLE
Change the way to inline method handle chains

### PR DESCRIPTION
**Pull request by cathyzhyi**
This change is part of the work to improve JSR292 performance. The old
way of inlining method handle chain is through multiple seperate
passes of targeted inlining opt group. This opt group inlines only
some special  methods related to JSR292. It relies on several
optimizations like local VP ,local CSE etc. to discover the next
method handles in the chain and request a new pass of the same
opt group if there are any. This strategy has 2 major problems:

1. When targeted inlining opts are done before the normal inlining like
before, only the method handle chains originated from the top level
methods are inlined. Simply swapping normal inlining and
targeted inlining in the opt strategies would solve this problem but
introducing new issues becuase the leaf java methods won't be inlined.

2. Relying on different passes of local opts to dicover the next method
handle in a method handle chain is expensive.

This change intends to fix those 2 problems by doing inlining of method
handle chain and normal inlining in one pass. There are 3 main parts:

1. RealEstimateCodeSize is changed to always do peeking for method handle
thunks to preserve the known object info through temps holding method
handle objects. This unlocks the next method handle without needing any
other optimizations.

2. Make sure method handle thunk are inlined.
   2.1 Change the method thunk body size to 0 and make them force inlining
target to make sure method thunks are always inlined for the entire
chain. Ideally, a fully optimized method thunk shouldn't contain any
real operations anyways. The leaf Java method at the end of method
handle chain would respect various heuristic in inliner as any other
java method and could be either inlined or not.

   2.2 In inlineCallTargets when chopping off the tailing targets because of
limited total budget, keep method handle thunks. This is intended to serve
the same function of the old targeted inlining which inlines a method
handle thunk whenever it's found and doesn't have same constraint as the
normal inlining.

   2.3 Check for remaining method handle invokes in an inlined target
and request for a pass of targeted inlining if there is any. This is
added due to the eager nature of inliner in case 2.1 and 2.2 together
still don't inline method handle that should be inlined.

4. Some recognized methods capsulated in isJSR292AlwaysWorthInlining are
added as hint to inliner heuristic that we favor inlining those methods.

Signed-off-by: Yi Zhang <yizhang@ca.ibm.com>

### CODE REVIEW

**Comment by cathyzhyi** [comment]
This change is still under functional and performance verification. I am opening this PR for discussions and first-round reviews. @andrewcraik @liqunl

**Comment by cathyzhyi** [comment]
this change is dependent on https://github.com/eclipse/openj9/pull/6175

**Comment by andrewcraik** [comment]
@cathyzhyi given the expense of the always do peeking operation in step 1 do we expect significant compile footprint increases or increases in compile time? Is there a plan to drop this requirement going forward.

I approve of the general design of allow the inliner to pick the targets to inline by looking through handle chains. I am not convinced that you can completely eliminate targeted inlining because of the eager nature of the current inliner and its sensitivity to block frequency. Is it not possible that the inliner will terminate without inlining all of the clean-up targets you want it to inline in step 3? Given the complexity of the inliner heuristics I would appreciate a detailed explanation on how you aim to eliminate all the handle invokes using 3.

I would appreciate a more complete description of item 4 - what is the new heuristic and what does it aim to do. What cases is it supposed to cover and where is the heuristic implemented?

**Comment by cathyzhyi** [comment]
@andrewcraik Thanks for raising the concerns. 
> @cathyzhyi given the expense of the always do peeking operation in step 1 do we expect significant compile footprint increases or increases in compile time? Is there a plan to drop this requirement going forward. 

Yes, peeking is expensive and takes around 30% of the entire footprint of inliner. And we expect the compilation time would increase as well. There is a follow up item to inline the method handle chain without peeking which is still only at prototyping state due to time constraint https://github.com/eclipse/openj9/issues/6204

> I am not convinced that you can completely eliminate targeted inlining because of the eager nature of the current inliner and its sensitivity to block frequency. Is it not possible that the inliner will terminate without inlining all of the clean-up targets you want it to inline in step 3

If inliner terminates because of the **caller** of mh thunk blowing the budget then even with targeted inlining we won't be able to see the thunk. As long as the caller is a potential inlining target, the new strategy would include the thunk because thunks are `forceInline` targets. There are various places in the code checks for `forceInline` to skip **exceedsSizeThreshould**(where the checks for frequency happens),  skip conditions in **doCorrectnessAndSizeChecksForInlineCallTarget** and in **weighCallSite**.  Those heuristic exists for `forceInline` targets already.  I only needed to add check for `forceInline` and set target size to 0 in **RealEstimateCodesize**. 

>Given the complexity of the inliner heuristics I would appreciate a detailed explanation on how you aim to eliminate all the handle invokes using 3.

The handle invokes are guaranteed to be inlined by a combination of step 2 and step 3. Step 2 guarantees thunk targets won't be removed because of thresholds for individual targets or a sub callgraph. Step 3 guarantee the targets won't be removed because of total budget on entire method (total weight and total node budget).

>I would appreciate a more complete description of item 4 - what is the new heuristic and what does it aim to do. What cases is it supposed to cover and where is the heuristic implemented?

In the old targeted inlining there is a combination of black and white list defined by https://github.com/eclipse/openj9/blob/ef6d8985c5b9fa5a15df5fe499dbc585c76afec9/runtime/compiler/env/VMJ9.cpp#L3100 to define whether a JSR292 related methods are inlineable. Those are not needed anymore because everything becomes inlineable now. Some recognized method capsulated in 
**isJSR292AlwaysWorthInlining** are added as hint to inliner heuristic that we favor inlining those methods. https://github.com/eclipse/openj9/pull/6173/files#diff-a4c456ab23834ff19a1bfb46122fa812R264 Those methods are added based on observations from profiling data of Nashorn and JRuby benchmarks.

**Comment by andrewcraik** [comment]
@cathyzhyi thanks for the responses.

Thanks for the explanation. It seems to me that we are going to have a savings by removing some of the additional opts in the targeted inlining pass as well as the extra inliner startups. Could you generate some performance numbers to show your throughput gain and total compile time with the change. If we have a nice gain (like you have described on a couple of the issues) and the compile-time is reasonable the cost of the peeking should be tolerable. We should still try to avoid the peeking, but the intermediate state with the peeking enabled will be tolerable for production use in my mind.

If you are convinced that 2 and 3 are sufficient could you add an optional walk of the trees in the inliner after inlining to verify all the handles have been inlined? You could do an extended test run with a fatal assert if handles are missed. My main concern is how robust this solution without any targeted inlining will be when the inliner is modified again in future. There are places where we do return early etc and having non-obvious performance implications from the inliner not doing the handles would be very unfortunate. Is the cost of having a pass to find and inline any straggling handles too high? At the very least the optional pass to verify that the handles are inlined would seem the minimal defense.

Thanks for making a note of the heuristic changes - they seem reasonable in light of all the other changes in this PR.

Overall, I think these are great changes and are something we should look to merge as soon as the builds are clean and you have the performance data to justify the change.

**Comment by cathyzhyi** [comment]
Measured  footprint and compilation time impact of the new inliner on JRuby. There was around 20%~50% increase in footprint and worse case 60% increase of compilation time.  Working on https://github.com/eclipse/openj9/issues/6204 to reduce impact on both.

**Comment by cathyzhyi** [comment]
Made some adjustment in the code to address this concern
> You could do an extended test run with a fatal assert if handles are missed. My main concern is how robust this solution without any targeted inlining will be when the inliner is modified again in future. There are places where we do return early etc and having non-obvious performance implications from the inliner not doing the handles would be very unfortunate. Is the cost of having a pass to find and inline any straggling handles too high? At the very least the optional pass to verify that the handles are inlined would seem the minimal defense.

A pass to find straggling handlers is added and if there is any a pass of targeted inlining will be requested. The extra compilation overhead shouldn't be a big issue as the walk is linear. 

This change is now dependent on https://github.com/eclipse/omr/pull/4146

**Comment by cathyzhyi** [comment]
this change is dependent on https://github.com/eclipse/openj9/pull/6791

**Comment by andrewcraik** [comment]
Jenkins test sanity all jdk8,jdk11

**Comment by andrewcraik** [comment]
Jenkins test sanity aix jdk11

**Comment by andrewcraik** [comment]
@liqunl review from you please as someone else who is working in the area.

File: runtime/compiler/optimizer/J9EstimateCodeSize.cpp [hunk 1]
```diff
@@ -1976,11 +2113,14 @@ TR_J9EstimateCodeSize::realEstimateCodeSize(TR_CallTarget *calltarget, TR_CallSt
             if (callSites[i]->numTargets()) //only add a callSite once, even though it may have more than one call target.
                {
                calltarget->addCallee(callSites[i]);
-               heuristicTrace(tracer(), "Depth %d: Subtracting %d from optimistic and real size to account for eliminating call", _recursionDepth, bci.estimatedCodeSize());
-               if (_optimisticSize > bci.estimatedCodeSize())
-                  _optimisticSize -= bci.estimatedCodeSize(); // subtract what we added before for the size of the call instruction
-               if (_realSize > bci.estimatedCodeSize())
-                  _realSize -= bci.estimatedCodeSize();
+               if (!callSites[i]->_initialCalleeMethod)
```
[the remark this answers is not in the document]
Comment (cathyzhyi) [reply runtime/compiler/optimizer/J9EstimateCodeSize.cpp]: Actually it's not needed. Deleted.

File: runtime/compiler/optimizer/J9EstimateCodeSize.cpp [hunk 2]
```diff
@@ -1266,6 +1323,81 @@ TR_J9EstimateCodeSize::realEstimateCodeSize(TR_CallTarget *calltarget, TR_CallSt
...
+                  TR_CallSite *callsite = new (comp()->trHeapMemory()) TR_J9MethodHandleCallSite( callStack._method, callNodeTreeTop,   parent,
+                                                                                    callNode, interfaceMethod, resolvedMethod->classOfMethod(),
+                                                                                    (int32_t) resolvedMethod->virtualCallSelector(cpIndex), cpIndex, resolvedMethod,
+                                                                                    resolvedSymbol, isIndirectCall, isInterface, newBCInfo, comp(),
+                                                                                    _recursionDepth, allconsts);
+
+                  TR_PrexArgInfo *argInfo = calltarget->_ecsPrexArgInfo;
+                  if (wasPeekingSuccessfull)
+                     {
+                     TR_PrexArgInfo::propagateReceiverInfoIfAvailable(methodSymbol, callsite, argInfo, tracer());
```
[the remark this answers is not in the document]
Comment (cathyzhyi) [reply runtime/compiler/optimizer/J9EstimateCodeSize.cpp]: Right, the receiver prex arg for thunk archetype is handled by `createPrexArgInfoForCallTarget` already. Shouldn't need to propagate from the caller. Deleted as suggested.

File: runtime/compiler/optimizer/InlinerTempForJ9.cpp [hunk 3]
```diff
@@ -4713,50 +4713,169 @@ TR_J9InlinerPolicy::validateArguments(TR_CallTarget *calltarget, TR_LinkHead<TR_
...
+
+   if (isJSR292SmallGetterMethod(resolvedMethod))
+      return true;
+
+   if (isJSR292SmallHelperMethod(resolvedMethod))
+      return true;
+
+   if (resolvedMethod->convertToMethod()->isArchetypeSpecimen())
+      return true;
+
```
[the remark this answers is not in the document]
Comment (cathyzhyi) [reply runtime/compiler/optimizer/InlinerTempForJ9.cpp]: Added as suggested.

File: runtime/compiler/optimizer/InlinerTempForJ9.cpp [hunk 4]
```diff
@@ -4713,50 +4713,169 @@ TR_J9InlinerPolicy::validateArguments(TR_CallTarget *calltarget, TR_LinkHead<TR_
...
+      return InlineableTarget;
+      }
+   else if (comp->getMethodHotness() >= warm)
+      {
+      // We are not an archetype specimen
+      // but because we're hot (or greater) we are allowed to inline JSR292 methods
+      return InlineableTarget;
+      }
+
+   //we are not an archetype specimen ourselves and we are warm or below, No inlining of JSR292 methods.
```
[the remark this answers is not in the document]
Comment (cathyzhyi) [reply runtime/compiler/optimizer/InlinerTempForJ9.cpp]: Changed the comment to correctly indicate what the code is doing. The intention is to do JSR292 inlining for warm and above.

File: runtime/compiler/optimizer/InlinerTempForJ9.cpp [hunk 5]
```diff
@@ -4713,50 +4713,169 @@ TR_J9InlinerPolicy::validateArguments(TR_CallTarget *calltarget, TR_LinkHead<TR_
...
-   return  static_cast<TR_InlinerFailureReason> (comp->fej9()->checkInlineableTarget(target, callsite, comp, false));
+   // for GPU, skip all the heuristic for JSR292 related methods
+   if (comp->hasIntStreamForEach())
+      return DontInline_Callee;
+
+   TR_ResolvedMethod * resolvedMethod = target->_calleeSymbol ? target->_calleeSymbol->getResolvedMethod():target->_calleeMethod;
+   if (!isJSR292Method(resolvedMethod))
+      return DontInline_Callee;
+
+   if (isJSR292AlwaysWorthInlining(resolvedMethod) || isJSR292LeafHandleInvoke(resolvedMethod))
```
[the remark this answers is not in the document]
Comment (cathyzhyi) [reply runtime/compiler/optimizer/InlinerTempForJ9.cpp]: Deleted `isJSR292LeafHandleInvoke` entirely as it's not needed anywhere.

File: runtime/compiler/optimizer/InlinerTempForJ9.cpp [hunk 6]
```diff
@@ -4713,50 +4713,169 @@ TR_J9InlinerPolicy::validateArguments(TR_CallTarget *calltarget, TR_LinkHead<TR_
...
+
+bool TR_J9InlinerPolicy::isJSR292LeafHandleInvoke(TR_ResolvedMethod *resolvedMethod)
+   {
+   TR::RecognizedMethod method =  resolvedMethod->getRecognizedMethod();
+   switch (method)
+      {
+      // single-level "leaf" handles
+      case TR::java_lang_invoke_DirectHandle_invokeExact:
+      case TR::java_lang_invoke_InterfaceHandle_invokeExact:
+      case TR::java_lang_invoke_VirtualHandle_invokeExact:
```
[the remark this answers is not in the document]
Comment (cathyzhyi) [reply runtime/compiler/optimizer/InlinerTempForJ9.cpp]: Thanks for pointing this out. Combining with other comment below. Deleted `isJSR292LeafHandleInvoke` entirely.

File: runtime/compiler/optimizer/InlinerTempForJ9.cpp [hunk 7]
```diff
@@ -4070,6 +4010,64 @@ void TR_MultipleCallTargetInliner::weighCallSite( TR_CallStack * callStack , TR_
...
+            generateNodeEstimate myEstimate;
+            recursivelyWalkCallTargetAndPerformAction(calltarget, myEstimate);
+            estimatedNumberOfNodes += myEstimate.getNodeEstimate();
+            /*
+             * ForceInline targets and JSR292 methods should always be inlined regarless of budget. However, with
+             * inlining methodhandle chain in normal inlining, the number of nodes can be tremendous resulting in
+             * compilations, especially those with higher opt level, eating up too much CPU time. The heuristic here
+             * is added to prevent compilations consuming too much CPU.
+             */
+            if (estimatedNumberOfNodes > 50000 && comp()->getMethodHotness() >= hot)
```
[the remark this answers is not in the document]
Comment (cathyzhyi) [reply runtime/compiler/optimizer/InlinerTempForJ9.cpp]: Added env var `TR_dontAbortCompilationEvenWithLargeInliningNodesEstimation` to guard the heuristic.

File: runtime/compiler/optimizer/EstimateCodeSize.cpp [hunk 8]
```diff
@@ -121,6 +121,9 @@ TR_EstimateCodeSize::isInlineable(TR_CallStack * prevCallStack, TR_CallSite *cal
    heuristicTrace(tracer(),"Depth %d: Created Call Site %p for call found at bc index %d. Signature %s  Looking for call targets.",
                              _recursionDepth, callsite, callsite->_byteCodeIndex, tracer()->traceSignature(callsite));
 
+   if (_inliner->getPolicy()->supressInliningRecognizedInitialCallee(callsite, _inliner->comp()))
```
[the remark this answers is not in the document]
Comment (cathyzhyi) [reply runtime/compiler/optimizer/EstimateCodeSize.cpp]: Added tracing msg.

**Comment by cathyzhyi** [comment]
@liqunl Can you take a look if all the concerns are addressed?

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win,plinux jdk8,jdk11

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win,plinux jdk8,jdk11

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win,plinux jdk8,jdk11

**Comment by pshipton** [comment]
jenkins test sanity xlinux,win,plinux jdk8,jdk11

**Comment by pshipton** [comment]
jenkins test sanity xlinux jdk8,jdk11

**Comment by pshipton** [comment]
jenkins test sanity xlinux,win,plinux jdk8,jdk11

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win,plinux jdk8,jdk11

**Comment by headius** [comment]
Hello there! I'm testing this out with JRuby a little bit (though it's nice to see others are too!) and I have some good news and some bad news.

First, the good news is that it appears to be much faster than the older build of OpenJ9 JDK11 I tested against.

Old J9:
```
openjdk version "11.0.2" 2019-01-15
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.2+9)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.12.1, JRE 11 Mac OS X amd64-64-Bit Compressed References 20190204_123 (JIT enabled, AOT enabled)
```

Old results on bench_mandelbrot.rb from https://github.com/headius/bench2018
```
$ jruby -Xcompile.invokedynamic ../bench2018/bench_mandelbrot.rb 
18.519094
17.313748
17.316665
17.08374
20.698155
20.069198
19.935275999999998
^C
```
(I bailed out since it wasn't getting better)

New results:
```
$ jruby -Xcompile.invokedynamic ../bench2018/bench_mandelbrot.rb 
17.404888
3.788252
15.950135
15.914788999999999
18.326113000000003
2.995778
3.027767
3.043336
3.060171
3.037857
3.0018269999999996
2.903241
2.919533
2.903168
3.0116910000000003
2.858603
2.8342460000000003
2.824973
2.991121
...
```

It's also faster than non-indy mode, though that indy warmup curve is pretty rough.

Old with non-indy mode:
```
$ jruby ../bench2018/bench_mandelbrot.rb 
5.842502
5.54749
5.153887
5.725445
5.522631
```

New with non-indy mode:
```
$ jruby ../bench2018/bench_mandelbrot.rb 
6.2411259999999995
4.676962
4.061717000000001
4.023565
3.5771
3.64104
3.494736
...
```

The bad news is that it still lags quite a bit behind Hotspot C2:
```
$ jruby -Xcompile.invokedynamic ../bench2018/bench_mandelbrot.rb 
1.058665
0.7813910000000001
0.781907
0.7761750000000001
0.7823680000000001
0.775401
0.790957
0.787907
```

And of course this is the sort of benchmark that Graal eats for breakfast (but I should also point out that we have *NOT* been able to get Graal to speed up any large apps yet):
```
$ pickjdk 3
New JDK: graalvm-ee-19.1.0

[] ~/projects/openj9-openjdk-jdk11 $ jruby -Xcompile.invokedynamic ../bench2018/bench_mandelbrot.rb 
0.144362
0.14385299999999998
0.142332
0.146192
0.14143199999999997
0.14179699999999998
0.14696399999999998
0.144637
0.142814
0.150497
0.141763
0.142508
...
```

Note that JRuby master (9.2.9) leverages more indy logic in non-indy mode, mostly for lazy construction of constant literals. That likely plays here in the non-indy mode numbers above.

**Comment by andrewcraik** [comment]
@headius thanks for the evaluation - @liqunl and others are still looking at the benchmarks. There is a change from @liqunl to read some final internal representation fields in the OpenJ9 MethodHandle representation which will likely provide a further boost (perhaps she can pass you patch to apply on top of @cathyzhyi 's change. There are also some object allocation overheads that are being investigated.

I think we will get near to the C2 level there is just some more tweaking and tuning to do on our JSR292 implementation since it isn't fully optimized yet (but is getting more so as you can see).

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win,plinux jdk8,jdk11

**Comment by liqunl** [comment]
@headius Glad to see that improvements show up in your build. We have changes pending to get us to a better position. With those changes, openj9's score is 2X of hotspot on mandelbrot. We have also identified more places in our MethodHandle implementation that we're going to improve.

The following changes may improve the performance a bit more
https://github.com/eclipse/openj9/pull/6112
https://github.com/liqunl/openj9/commit/2c74f19ccf58844f70a373abc972a0b4afe969de

**Comment by headius** [comment]
@liqunl Thank you for the additional links!

I can confirm they do seem to help, mostly on the warmup side. The overall peak numbers seem like they're probably about the same range as running off this PR alone:

```
$ jruby -Xcompile.invokedynamic ../bench2018/bench_mandelbrot.rb 
5.06217
4.252303
4.226223
3.0020789999999997
2.948776
2.927709
2.945977
2.971966
3.033959
2.932014
3.027892
2.9487929999999998
2.977682
3.147387
2.9171479999999996
2.9327159999999997
3.045137
2.97194
2.953372
2.931979
```

Note there are no iterations in the high teens, where there were several before.

**Comment by headius** [comment]
Big improvement on red/black benchmark!

OLD
```
$ jruby -Xcompile.invokedynamic ../bench2018/bench_red_black.rb
12.684894
10.600425999999999
10.699691
10.590059
...
```

NEW
```
$ jruby -Xcompile.invokedynamic ../bench2018/bench_red_black.rb 
6.467919
1.677888
1.5920150000000002
1.658796
1.6530960000000001
1.435051
1.684435
2.290969
1.3159690000000002
1.194621
1.202193
1.237888
1.212079
1.167899
1.3245989999999999
1.152364
1.126513
1.082473
1.203363
1.389172
```

**Comment by andrewcraik** [comment]
Windows fail is infra sanity is clean.

**Comment by andrewcraik** [comment]
Perf looks good for the 292 per the above - merging.

**Comment by DanHeidinga** [comment]
Congrats @cathyzhyi on getting this major PR landed!  A big thank you to you and all involved in getting this (and related 292) improvements into OpenJ9

**Pull request merged**